src/VisualStudio/CSharp/Impl/LanguageService/CSharpCodeCleanupFixer.cs

```diff
@@ -0,0 +1,348 @@
+        [Name(CSharpRemoveUnusedVariableCodeFixProvider.CS0168)]
+        [Order(After = IDEDiagnosticIds.RemoveUnnecessaryCastDiagnosticId)]
+        [ConfigurationKey("unused")]
+        [HelpLink("https://www.microsoft.com")]
+        [LocalizedName(typeof(CSharpFeaturesResources), nameof(CSharpFeaturesResources.Remove_unused_variables))]
+        public static readonly FixIdDefinition CS0168;
+
+        [Export]
+        [FixId(CSharpRemoveUnusedVariableCodeFixProvider.CS0219)]
+        [Name(CSharpRemoveUnusedVariableCodeFixProvider.CS0219)]
+        [Order(After = IDEDiagnosticIds.RemoveUnnecessaryCastDiagnosticId)]
+        [ConfigurationKey("unused")]
+        [HelpLink("https://www.microsoft.com")]
+        [LocalizedName(typeof(CSharpFeaturesResources), nameof(CSharpFeaturesResources.Remove_unused_variables))]
+        public static readonly FixIdDefinition CS0219;
+
+        [Export]
+        [FixId(IDEDiagnosticIds.UseObjectInitializerDiagnosticId)]
+        [Name(IDEDiagnosticIds.UseObjectInitializerDiagnosticId)]
+        [Order(After = IDEDiagnosticIds.PreferBuiltInOrFrameworkTypeDiagnosticId)]
+        [ConfigurationKey("unused")]
+        [HelpLink("https://www.microsoft.com")]
+        [LocalizedName(typeof(CSharpFeaturesResources), nameof(CSharpFeaturesResources.Apply_object_collection_initialization_preferences))]
+        public static readonly FixIdDefinition UseObjectInitializerDiagnosticId;
+
+        [Export]
+        [FixId(IDEDiagnosticIds.UseCollectionInitializerDiagnosticId)]
+        [Name(IDEDiagnosticIds.UseCollectionInitializerDiagnosticId)]
+        [Order(After = IDEDiagnosticIds.PreferBuiltInOrFrameworkTypeDiagnosticId)]
+        [ConfigurationKey("unused")]
+        [HelpLink("https://www.microsoft.com")]
+        [LocalizedName(typeof(CSharpFeaturesResources), nameof(CSharpFeaturesResources.Apply_object_collection_initialization_preferences))]
+        public static readonly FixIdDefinition UseCollectionInitializerDiagnosticId;
+
+        [Export]
+        [FixId(RemoveUnusedImportsFixId)]
+        [Name(RemoveUnusedImportsFixId)]
+        [ConfigurationKey("unused")]
+        [HelpLink("https://www.microsoft.com")]
+        [ExportMetadata("EnableByDefault", true)]
+        [LocalizedName(typeof(CSharpVSResources), nameof(CSharpVSResources.Remove_unnecessary_usings))]
+        public static readonly FixIdDefinition RemoveUnusedImports;
+
+        [Export]
+        [FixId(SortImportsFixId)]
+        [Name(SortImportsFixId)]
+        [Order(After = RemoveUnusedImportsFixId)]
+        [ConfigurationKey("unused")]
+        [HelpLink("https://www.microsoft.com")]
+        [ExportMetadata("EnableByDefault", true)]
+        [LocalizedName(typeof(CSharpVSResources), nameof(CSharpVSResources.Sort_usings))]
+        public static readonly FixIdDefinition SortImports;
+
+        public override Task<bool> FixAsync(ICodeCleanUpScope scope, ICodeCleanUpExecutionContext context, CancellationToken cancellationToken)
```

Comment:
what does the caller do with this result?

Reply:
According to the documentation:

> bool if the fixer succeeded false otherwise

Reply:
@olegtk can you provide more information on this?  thanks!

Reply:
looking at the current implementation we don't really use return value. I think the intent is to undo any changes if a fixer fails, which should probably include both a fixer returning false and throwing an exception. @crmann1 ?

Reply:
> looking at the current implementation we don't really use return value. I think the intent is to undo any changes if a fixer fails, which should probably include both a fixer returning false and throwing an exception. @crmann1 ?

That seems like super weird behavior :)   Do we have any precedent for that elsewhere?  How do you guys track changes, or know what to undo?  Do you undo changes to projects/solutions/etc.?